Upgrade to Apache Calcite 1.21

amitvc opened the pull request:
WIP -  DRAFT

eolivelli commented on herddb-core/src/main/java/herddb/sql/CalciteEnumUtils.java:
```diff
@@ -0,0 +1,26 @@
+
+public class CalciteEnumUtils {
+
+    public static JoinType toLinq4jJoinType(JoinRelType joinRelType) {
```
We should add an unit test with a query for every type of join.
You can add a breakpoint or a throw new RuntimeException() on every case clause and then run the tests about joins, this way you will see if there is some kind of join not covered by tests

amitvc replied:
Sounds good. I will be adding these tests shortly.

amitvc replied:
I will be adding UT for this tonight.

eolivelli commented:
In order to make travis run again please change from oraclejdk11 to openjdk11 in travis.yml file in the root directory

aluccaroni commented:
I don't know if this is the correct place or, maybe, if it's better to do it in a new pull but we have alto to take a look at these issue resolved in Calcite 1.18:
https://issues.apache.org/jira/browse/CALCITE-2662
https://issues.apache.org/jira/browse/CALCITE-2591
https://issues.apache.org/jira/browse/CALCITE-2405

amitvc commented:
Ran into a few more issues. Plan on working on this tonight.

eolivelli commented:
Did you push all of your changes?
I can start investigating your problem if this your current work

amitvc commented:
Yes, the code is current. I am going to comment out that test and see if others break too

amitvc commented:
A couple of more observations - 
The following query breaks - 
"UPDATE tblspace1.table1 set n1=1000"
                    + "WHERE k1 in (SELECT fk FROM tblspace1.table2 WHERE k2=?)"
https://github.com/amitvc/herddb/blob/master/herddb-core/src/test/java/herddb/core/SimpleSubqueryTest.java#L225

When the CalcitePlanner runs and recursively tries to evaluate the query and validate it one of the SQLNodes for the where part of the query has the following value - 
**`K1` IN (SELECT `table2`.`fk` AS `FK`
FROM `tblspace1`.`table2` AS `TABLE2`
WHERE `table2`.`k2` = ?)**
But the SqlValidatorImpl::nodeToTypeMap does not contain a value that matches this type. The reason is calcite is appending Table prefix to K1. That happens DelagatingScope::fullyQualify function. The node value in the nodeToTypeMap is **"`table1`.`k1` IN (SELECT `table2`.`fk` AS `FK`
FROM `tblspace1`.`table2` AS `TABLE2`
WHERE `table2`.`k2` = ?)" -> "BOOLEAN"**
You can see that the table namespace has been attached to the column name. I wonder if this is some configuration setting we can add when we bootstrap the CalcitePlanner. But this will most likely break a lot of our tests and code in production if Calcite is expecting column names are prefixed with table name in queries like this.

amitvc commented:
I wonder if we need to configure SqlToRelConverter when we bootstrap FrameworkConfig

eolivelli commented:
@amitvx can you please paste here the error about min/max aggregation?
I have sent an email to Calcite ML.
We have to dig into our issue and tell if it is a bug in HerdDB or a regression in Calcite

eolivelli commented:
About the UPDATE query with the subquery, can you also paste here the full error? Is it in a test?

amitvc commented:
The query is - UPDATE tblspace1.table1 set n1=1000WHERE k1 in (SELECT fk FROM tblspace1.table2 WHERE k2=?)

Here is the exception in the update query - 


java.lang.UnsupportedOperationException: class org.apache.calcite.sql.SqlBasicCall: `K1` IN (SELECT `table2`.`fk` AS `FK`
FROM `tblspace1`.`table2` AS `TABLE2`
WHERE `table2`.`k2` = ?)

	at org.apache.calcite.util.Util.needToImplement(Util.java:967)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.getValidatedNodeType(SqlValidatorImpl.java:1579)
	at org.apache.calcite.sql2rel.SqlToRelConverter.findSubQueries(SqlToRelConverter.java:1802)
	at org.apache.calcite.sql2rel.SqlToRelConverter.findSubQueries(SqlToRelConverter.java:1776)
	at org.apache.calcite.sql2rel.SqlToRelConverter.replaceSubQueries(SqlToRelConverter.java:1011)
	at org.apache.calcite.sql2rel.SqlToRelConverter.convertUpdate(SqlToRelConverter.java:3570)
	at org.apache.calcite.sql2rel.SqlToRelConverter.convertQueryRecursive(SqlToRelConverter.java:3172)
	at org.apache.calcite.sql2rel.SqlToRelConverter.convertQuery(SqlToRelConverter.java:563)
	at org.apache.calcite.prepare.PlannerImpl.rel(PlannerImpl.java:254)
	at herddb.sql.CalcitePlanner.runPlanner(CalcitePlanner.java:506)
	at herddb.sql.CalcitePlanner.translate(CalcitePlanner.java:293)
	at herddb.core.TestUtils.executeUpdate(TestUtils.java:43)
	at herddb.core.SimpleSubqueryTest.deleteWithSubQueryTest(SimpleSubqueryTest.java:224)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)

amitvc commented:
Hi Enrico,

The query is - UPDATE tblspace1.table1 set n1=1000WHERE k1 in (SELECT fk
FROM tblspace1.table2 WHERE k2=?)

Here is the exception in the update query -

java.lang.UnsupportedOperationException: class
org.apache.calcite.sql.SqlBasicCall: K1 IN (SELECT table2.fk AS FK
FROM tblspace1.table2 AS TABLE2
WHERE table2.k2 = ?)

at org.apache.calcite.util.Util.needToImplement(Util.java:967)
at org.apache.calcite.sql.validate.SqlValidatorImpl.getValidatedNodeType(SqlValidatorImpl.java:1579)
at org.apache.calcite.sql2rel.SqlToRelConverter.findSubQueries(SqlToRelConverter.java:1802)
at org.apache.calcite.sql2rel.SqlToRelConverter.findSubQueries(SqlToRelConverter.java:1776)
at org.apache.calcite.sql2rel.SqlToRelConverter.replaceSubQueries(SqlToRelConverter.java:1011)
at org.apache.calcite.sql2rel.SqlToRelConverter.convertUpdate(SqlToRelConverter.java:3570)
at org.apache.calcite.sql2rel.SqlToRelConverter.convertQueryRecursive(SqlToRelConverter.java:3172)
at org.apache.calcite.sql2rel.SqlToRelConverter.convertQuery(SqlToRelConverter.java:563)
at org.apache.calcite.prepare.PlannerImpl.rel(PlannerImpl.java:254)
at herddb.sql.CalcitePlanner.runPlanner(CalcitePlanner.java:506)
at herddb.sql.CalcitePlanner.translate(CalcitePlanner.java:293)
at herddb.core.TestUtils.executeUpdate(TestUtils.java:43)
at herddb.core.SimpleSubqueryTest.deleteWithSubQueryTest(SimpleSubqueryTest.java:224)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:497)
at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)


On Sun, Jul 21, 2019 at 12:50 PM Enrico Olivelli <notifications@github.com>
wrote:

> About the UPDATE query with the subquery, can you also paste here the full
> error? Is it in a test?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/diennea/herddb/pull/404?email_source=notifications&email_token=ABH7RV7CK36WSNKNOMUJ5MTQAS4YBA5CNFSM4H4CC4F2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2OKLNY#issuecomment-513582519>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABH7RV74BGP74XH6YA7A2ZLQAS4YBANCNFSM4H4CC4FQ>
> .
>

eolivelli commented:
Did you have time to take another look?
We could try to upgrade to 1.19 and see if it is a regression of 1.20, or maybe it is better to start by upgrading to 1.18.

I remember I tried to move to 1.18 without issues when it was released

amitvc commented:
I can try to upgrade to 1.19. I will have a PR out tonight. Have you got
any response from the Calcite community about the errors we are getting
with 1.20 ?

On Wed, Jul 24, 2019 at 7:49 AM Enrico Olivelli <notifications@github.com>
wrote:

> Did you have time to take another look?
> We could try to upgrade to 1.19 and see if it is a regression of 1.20, or
> maybe it is better to start by upgrading to 1.18.
>
> I remember I tried to move to 1.18 without issues when it was released
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/diennea/herddb/pull/404?email_source=notifications&email_token=ABH7RV4NZQ7ULO6OUSPQGETQBBTWHA5CNFSM4H4CC4F2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2WSQJY#issuecomment-514664487>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABH7RV7IQKBEE6TFQUVC7ETQBBTWHANCNFSM4H4CC4FQ>
> .
>

eolivelli commented:
Il giorno mer 24 lug 2019 alle ore 17:27 Amit Chavan <
notifications@github.com> ha scritto:

> I can try to upgrade to 1.19. I will have a PR out tonight. Have you got
> any response from the Calcite community about the errors we are getting
> with 1.20 ?
>

No.
I did not file a bug. I have only sent an email to dev@calcite.apache.org
maybe you can subscribe to that list


>
> On Wed, Jul 24, 2019 at 7:49 AM Enrico Olivelli <notifications@github.com>
> wrote:
>
> > Did you have time to take another look?
> > We could try to upgrade to 1.19 and see if it is a regression of 1.20, or
> > maybe it is better to start by upgrading to 1.18.
> >
> > I remember I tried to move to 1.18 without issues when it was released
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/diennea/herddb/pull/404?email_source=notifications&email_token=ABH7RV4NZQ7ULO6OUSPQGETQBBTWHA5CNFSM4H4CC4F2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2WSQJY#issuecomment-514664487
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/ABH7RV7IQKBEE6TFQUVC7ETQBBTWHANCNFSM4H4CC4FQ
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/diennea/herddb/pull/404?email_source=notifications&email_token=ACIHZNVQHFCSTYVBTHAG4I3QBBYH5A5CNFSM4H4CC4F2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2WWP2Y#issuecomment-514680811>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACIHZNWPGIOHRY3A5GDLPK3QBBYH5ANCNFSM4H4CC4FQ>
> .
>

amitvc commented:
Hi Enrico,

I upgraded herddb to calcite 1.19.0. There were no breaking changes and all
tests passed. So upgrading to 1.20.0 obviously has API changes along with
the issues we encountered that I mentioned in my previous email. Should we
just upgrade to 1.19.0 and perform more regression for now and open up
discussions with Calcite community about the other issues?

Thanks,
Amit

On Wed, Jul 24, 2019 at 8:29 AM Enrico Olivelli <notifications@github.com>
wrote:

> Il giorno mer 24 lug 2019 alle ore 17:27 Amit Chavan <
> notifications@github.com> ha scritto:
>
> > I can try to upgrade to 1.19. I will have a PR out tonight. Have you got
> > any response from the Calcite community about the errors we are getting
> > with 1.20 ?
> >
>
> No.
> I did not file a bug. I have only sent an email to dev@calcite.apache.org
> maybe you can subscribe to that list
>
>
> >
> > On Wed, Jul 24, 2019 at 7:49 AM Enrico Olivelli <
> notifications@github.com>
> > wrote:
> >
> > > Did you have time to take another look?
> > > We could try to upgrade to 1.19 and see if it is a regression of 1.20,
> or
> > > maybe it is better to start by upgrading to 1.18.
> > >
> > > I remember I tried to move to 1.18 without issues when it was released
> > >
> > > —
> > > You are receiving this because you authored the thread.
> > > Reply to this email directly, view it on GitHub
> > > <
> >
> https://github.com/diennea/herddb/pull/404?email_source=notifications&email_token=ABH7RV4NZQ7ULO6OUSPQGETQBBTWHA5CNFSM4H4CC4F2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2WSQJY#issuecomment-514664487
> > >,
> > > or mute the thread
> > > <
> >
> https://github.com/notifications/unsubscribe-auth/ABH7RV7IQKBEE6TFQUVC7ETQBBTWHANCNFSM4H4CC4FQ
> > >
> > > .
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/diennea/herddb/pull/404?email_source=notifications&email_token=ACIHZNVQHFCSTYVBTHAG4I3QBBYH5A5CNFSM4H4CC4F2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2WWP2Y#issuecomment-514680811
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/ACIHZNWPGIOHRY3A5GDLPK3QBBYH5ANCNFSM4H4CC4FQ
> >
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/diennea/herddb/pull/404?email_source=notifications&email_token=ABH7RV64MUJSRORKTF4NTSTQBBYPLA5CNFSM4H4CC4F2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2WWWXQ#issuecomment-514681694>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABH7RV2PU5JRLBBXUMGA7ILQBBYPLANCNFSM4H4CC4FQ>
> .
>

eolivelli commented:
Cool.
Please create a new PR with only the switch to 1.19.
It is better to move one step at a time

amitvc commented:
Will do.

On Thu, Jul 25, 2019 at 8:18 AM Enrico Olivelli <notifications@github.com>
wrote:

> Cool.
> Please create a new PR with only the switch to 1.19.
> It is better to move one step at a time
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/diennea/herddb/pull/404?email_source=notifications&email_token=ABH7RV3F65OI3EQ2REBL6GTQBG75PA5CNFSM4H4CC4F2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2ZZWZI#issuecomment-515087205>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABH7RV5KXTDGFRBQ4XMMD3DQBG75PANCNFSM4H4CC4FQ>
> .
>

amitvc commented:
https://github.com/diennea/herddb/pull/420 for calcite 1.19

On Thu, Jul 25, 2019 at 9:34 AM Amit Chavan <achavan1@gmail.com> wrote:

> Will do.
>
> On Thu, Jul 25, 2019 at 8:18 AM Enrico Olivelli <notifications@github.com>
> wrote:
>
>> Cool.
>> Please create a new PR with only the switch to 1.19.
>> It is better to move one step at a time
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/diennea/herddb/pull/404?email_source=notifications&email_token=ABH7RV3F65OI3EQ2REBL6GTQBG75PA5CNFSM4H4CC4F2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2ZZWZI#issuecomment-515087205>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABH7RV5KXTDGFRBQ4XMMD3DQBG75PANCNFSM4H4CC4FQ>
>> .
>>
>

eolivelli commented:
@amitvc as we see that 1.19 is good we can take a look to the changes in 1.20 and see that happened.
I will have time in the week end.
Please rebase this branch to latest master as soon as we merge #420

amitvc commented:
@eolivelli I rebased from the latest master.

eolivelli commented:
Great.
Currently @diegosalvi is working on a bad problem on secondary indexes, then we will cut another release.
I will take a look to this problem soon.
In the meanwhile you could pick up some other ticket.
I would like to have support for XAResource, this will enable full support for JavaEE+JakartaEE and it will be a great feature. 
Otherwise feel free to pick up any of the other tasks

amitvc commented:
Do you have a ticket written up for XAResource and what we would want to
do? I can work on it. I will read about XAResource tonight and look for a
ticket in the morning to see what needs to be done.

On Mon, Jul 29, 2019 at 9:36 PM Enrico Olivelli <notifications@github.com>
wrote:

> Great.
> Currently @diegosalvi <https://github.com/diegosalvi> is working on a bad
> problem on secondary indexes, then we will cut another release.
> I will take a look to this problem soon.
> In the meanwhile you could pick up some other ticket.
> I would like to have support for XAResource, this will enable full support
> for JavaEE+JakartaEE and it will be a great feature.
> Otherwise feel free to pick up any of the other tasks
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/diennea/herddb/pull/404?email_source=notifications&email_token=ABH7RV77UVIB2B55BAHTGRDQB7AMDA5CNFSM4H4CC4F2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3CXN4Y#issuecomment-516257523>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABH7RV7FNXJX37QRDE6GAJ3QB7AMDANCNFSM4H4CC4FQ>
> .
>

eolivelli commented:
@amitvc I have rebased your branch to current master and run tests.
All of the failure appear to me as regressions in Calcite.
I am saying that because the error is in the SQL Planner, and HerdDB starts working after Calcite work, the only input to Calcite is about configuration and Table Schemas.

I will report here the errors, but I will have to report them to Calcite dev list, open issues and hopefully find some help.

In the meanwhile I suggest you to subscribe to Apache Calcite mailing lists.

We will also have to reproduce the failures inside Calcite tests corpus, this way they will be able to work on the problems.
Maybe you will be the one who fixes them. 
I have already contributed to Calcite, it is a very good community

eolivelli commented:
Failed tests:

`[ERROR]   RawSQLTest.simpleMinMaxTest:1389 » Runtime Error while applying rule ProjectSc...`

The query is:
`SELECT MIN(n1) as mi, MAX(n1) as ma FROM tblspace1.tsql`

Error:
```
java.lang.RuntimeException: Error while applying rule ProjectScanRule:interpreter, args [rel#264:EnumerableProject.ENUMERABLE.[](input=RelSubset#263,MI=$1,MA=$1), rel#317:EnumerableInterpreter.ENUMERABLE.[](input=RelSubset#291), rel#290:BindableTableScan.BINDABLE.[](table=[tblspace1, tsql],filters=[=($0, 'no_results')])]
	at org.apache.calcite.plan.volcano.VolcanoRuleCall.onMatch(VolcanoRuleCall.java:235)
	at org.apache.calcite.plan.volcano.VolcanoPlanner.findBestExp(VolcanoPlanner.java:631)
	at herddb.sql.CalcitePlanner.runPlanner(CalcitePlanner.java:527)
	at herddb.sql.CalcitePlanner.translate(CalcitePlanner.java:293)
	at herddb.core.TestUtils.scan(TestUtils.java:70)
	at herddb.core.RawSQLTest.simpleMinMaxTest(RawSQLTest.java:1389)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
Caused by: java.lang.IllegalArgumentException: source #1 is already mapped to target #1
	at org.apache.calcite.util.mapping.Mappings$SurjectionWithInverse.set(Mappings.java:1326)
	at org.apache.calcite.rel.core.Project.getMapping(Project.java:279)
	at org.apache.calcite.rel.core.Project.getMapping(Project.java:250)
	at org.apache.calcite.rel.rules.ProjectTableScanRule.apply(ProjectTableScanRule.java:107)
	at org.apache.calcite.rel.rules.ProjectTableScanRule$2.onMatch(ProjectTableScanRule.java:83)
	at org.apache.calcite.plan.volcano.VolcanoRuleCall.onMatch(VolcanoRuleCall.java:208)
	... 30 more
```

test:
`[ERROR]   SimpleSubqueryTest.deleteWithSubQueryTest:224 » UnsupportedOperation class org...`

query:
```
UPDATE tblspace1.table1 set n1=1000"
                    + "WHERE k1 in (SELECT fk FROM tblspace1.table2 WHERE k2=?)
```

error:
```
class org.apache.calcite.sql.SqlBasicCall: `K1` IN (SELECT `table2`.`fk` AS `FK`
FROM `tblspace1`.`table2` AS `TABLE2`
WHERE `table2`.`k2` = ?)
	at org.apache.calcite.util.Util.needToImplement(Util.java:967)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.getValidatedNodeType(SqlValidatorImpl.java:1579)
	at org.apache.calcite.sql2rel.SqlToRelConverter.findSubQueries(SqlToRelConverter.java:1802)
	at org.apache.calcite.sql2rel.SqlToRelConverter.findSubQueries(SqlToRelConverter.java:1776)
	at org.apache.calcite.sql2rel.SqlToRelConverter.replaceSubQueries(SqlToRelConverter.java:1011)
	at org.apache.calcite.sql2rel.SqlToRelConverter.convertUpdate(SqlToRelConverter.java:3570)
	at org.apache.calcite.sql2rel.SqlToRelConverter.convertQueryRecursive(SqlToRelConverter.java:3172)
	at org.apache.calcite.sql2rel.SqlToRelConverter.convertQuery(SqlToRelConverter.java:563)
	at org.apache.calcite.prepare.PlannerImpl.rel(PlannerImpl.java:254)
	at herddb.sql.CalcitePlanner.runPlanner(CalcitePlanner.java:506)
	at herddb.sql.CalcitePlanner.translate(CalcitePlanner.java:293)
	at herddb.core.TestUtils.executeUpdate(TestUtils.java:43)
```

test:
`[ERROR]   CalcitePlannerTest.simplePlansTests:108->plan:763 » UnsupportedOperation class...`

query:
`update tblspace1.tsql set n1=? where n1 in (select b.n1*2 from tblspace1.tsql b)`

error:
```
java.lang.UnsupportedOperationException: 
class org.apache.calcite.sql.SqlBasicCall: `N1` IN (SELECT `B`.`n1` * 2
FROM `tblspace1`.`tsql` AS `B`)
	at org.apache.calcite.util.Util.needToImplement(Util.java:967)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.getValidatedNodeType(SqlValidatorImpl.java:1579)
	at org.apache.calcite.sql2rel.SqlToRelConverter.findSubQueries(SqlToRelConverter.java:1802)
	at org.apache.calcite.sql2rel.SqlToRelConverter.findSubQueries(SqlToRelConverter.java:1776)
	at org.apache.calcite.sql2rel.SqlToRelConverter.replaceSubQueries(SqlToRelConverter.java:1011)
	at org.apache.calcite.sql2rel.SqlToRelConverter.convertUpdate(SqlToRelConverter.java:3570)
	at org.apache.calcite.sql2rel.SqlToRelConverter.convertQueryRecursive(SqlToRelConverter.java:3172)
	at org.apache.calcite.sql2rel.SqlToRelConverter.convertQuery(SqlToRelConverter.java:563)
	at org.apache.calcite.prepare.PlannerImpl.rel(PlannerImpl.java:254)
	at herddb.sql.CalcitePlanner.runPlanner(CalcitePlanner.java:506)
	at herddb.sql.CalcitePlanner.translate(CalcitePlanner.java:293)
	at herddb.sql.CalcitePlannerTest.plan(CalcitePlannerTest.java:763)
	at herddb.sql.CalcitePlannerTest.dirtySQLPlansTests(CalcitePlannerTest.java:216)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at 
```

amitvc commented:
@Enrico Olivelli <eolivelli@gmail.com>
When I upgraded to 1.20 I experienced the same behavior. All exceptions
were happening inside calcite code. Initially I thought we need to pass
some new configuration when we are bootstrapping calcite but so far I have
not been able to figure that out. I will subscribe to apache calcite
mailing lists and report the errors to the calcite community. I can take
this off your plate and keep you updated as to what I find out from calcite
community.

On Wed, Jul 31, 2019 at 5:39 AM Enrico Olivelli <notifications@github.com>
wrote:

> @amitvc <https://github.com/amitvc> I have rebased your branch to current
> master and run tests.
> All of the failure appear to me as regressions in Calcite.
> I am saying that because the error is in the SQL Planner, and HerdDB
> starts working after Calcite work, the only input to Calcite is about
> configuration and Table Schemas.
>
> I will report here the errors, but I will have to report them to Calcite
> dev list, open issues and hopefully find some help.
>
> In the meanwhile I suggest you to subscribe to Apache Calcite mailing
> lists.
>
> We will also have to reproduce the failures inside Calcite tests corpus,
> this way they will be able to work on the problems.
> Maybe you will be the one who fixes them.
> I have already contributed to Calcite, it is a very good community
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/diennea/herddb/pull/404?email_source=notifications&email_token=ABH7RVYW2UCLSNPS5IMISLTQCGBXPA5CNFSM4H4CC4F2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3HDMCA#issuecomment-516830728>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABH7RV62W2X3UVYDHLSHLATQCGBXPANCNFSM4H4CC4FQ>
> .
>

eolivelli commented:
@amitvc in order to get more support from Calcite team it is better to move to 1.21.0-SNAPSHOT

currently I have these problems:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project herddb-core: Compilation failure: Compilation failure: 
[ERROR] /home/diennea.lan/enrico.olivelli/dev/herddb/herddb-core/src/main/java/herddb/sql/CalcitePlanner.java:[64,45] cannot find symbol
[ERROR]   symbol:   class EnumerableSemiJoin
[ERROR]   location: package org.apache.calcite.adapter.enumerable
[ERROR] /home/diennea.lan/enrico.olivelli/dev/herddb/herddb-core/src/main/java/herddb/sql/CalcitePlanner.java:[873,46] cannot find symbol
[ERROR]   symbol:   class EnumerableSemiJoin
[ERROR]   location: class herddb.sql.CalcitePlanner
[ERROR] /home/diennea.lan/enrico.olivelli/dev/herddb/herddb-core/src/main/java/herddb/sql/CalcitePlanner.java:[578,36] cannot find symbol
[ERROR]   symbol:   class EnumerableSemiJoin
[ERROR]   location: class herddb.sql.CalcitePlanner
[ERROR] /home/diennea.lan/enrico.olivelli/dev/herddb/herddb-core/src/main/java/herddb/sql/CalcitePlanner.java:[579,13] cannot find symbol
[ERROR]   symbol:   class EnumerableSemiJoin
[ERROR]   location: class herddb.sql.CalcitePlanner
[ERROR] /home/diennea.lan/enrico.olivelli/dev/herddb/herddb-core/src/main/java/herddb/sql/CalcitePlanner.java:[579,40] cannot find symbol
[ERROR]   symbol:   class EnumerableSemiJoin
[ERROR]   location: class herddb.sql.CalcitePlanner

```

this way we can test their master and provide a better feedback

amitvc commented:
I will change to use 1.21.0-SNAPSHOT and push the change tonight. Also have the Calcite team looking at the issues we reported and provided them with Unit tests to reproduce the issue.

eolivelli commented:
Very good.
Thank you

eolivelli commented:
@amitvc can you please link here the JIRAs and the Pull requests of the fixes from Calcite JIRA and Calcite GitHub repository ?

amitvc commented:
Calcite team has opened up tickets to address the issues we identified with 1.20 release of calcite.

https://issues.apache.org/jira/browse/CALCITE-3228
PR for this - 
https://github.com/apache/calcite/pull/1355

chunweilei commented:
> The select min/max query now works but the update query still fails with the same error

Please see [PR#1358](https://github.com/apache/calcite/pull/1358) for the update query.

eolivelli commented:
@amitvc now that Calcite fixes have been merged you can update this PR to use Calcite SNAPSHOT version
This way we can review and be ready to merge as soon Calcite is released.

Thank you @chunweilei  and to the ApacheCalcite community!

amitvc commented:
Will do @Enrico Olivelli <eolivelli@gmail.com>
Thank you @chunweilei for getting the fixes in.

On Sun, Sep 1, 2019 at 4:05 AM Enrico Olivelli <notifications@github.com>
wrote:

> @amitvc <https://github.com/amitvc> now that Calcite fixes have been
> merged you can update this PR to use Calcite SNAPSHOT version
> This way we can review and be ready to merge as soon Calcite is released.
>
> Thank you @chunweilei <https://github.com/chunweilei> and to the
> ApacheCalcite community!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/diennea/herddb/pull/404?email_source=notifications&email_token=ABH7RV65SCGHD7WABH2YVRLQHOOWXA5CNFSM4H4CC4F2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5T737Y#issuecomment-526908927>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABH7RV6YJ3HLDE54NLMLSILQHOOWXANCNFSM4H4CC4FQ>
> .
>

amitvc commented:
@eolivelli I am now working on lastest calcite master which I locally build [1.21.0-SNAPSHOT].

eolivelli commented:
Very good
I think that Calcite community deploys nightly snapshots to a shared repo.

We can use that repo in order to be able to build this branch on Travis.

I will take a look tomorrow, please push your changes so that we can check them out

amitvc commented:
Looks like EnumerableSemiJoin has been removed in 1.21 calcite and replaced with EnumerableNestedLoopJoin. I am looking to see how that will affect herddb since we need left and right keys SemiJoinOp.

eolivelli commented:
You can push your current work, this way I can check

eolivelli commented:
@amitvc please take a look here https://github.com/apache/calcite/pull/1289

amitvc commented:
@eolivelli Pushing some changes in next hour

amitvc commented:
@eolivelli I ran the test suite against my changes. The test that failed first is 
CalcitePlanner::joinsPlansTest. Here is the exception
java.lang.ArrayIndexOutOfBoundsException: 3
	at herddb.model.Table.isPrimaryKeyColumn(Table.java:145)
	at herddb.codec.DataAccessorForFullRecord.fieldEqualsTo(DataAccessorForFullRecord.java:74)
	at herddb.sql.expressions.AccessCurrentRowExpression.opEqualsTo(AccessCurrentRowExpression.java:48)
	at herddb.sql.expressions.CompiledNotEqualsExpression.evaluate(CompiledNotEqualsExpression.java:40)
	at herddb.model.planner.FilterOp$FilteredDataScanner.fetchNext(FilterOp.java:113)
	at herddb.model.planner.FilterOp$FilteredDataScanner.<init>(FilterOp.java:103)
	at herddb.model.planner.FilterOp.execute(FilterOp.java:84)
	at herddb.model.planner.ProjectOp.execute(ProjectOp.java:182)
	at herddb.model.planner.PlannerOp.executeAsync(PlannerOp.java:70)
	at herddb.core.TableSpaceManager.executePlannedOperationStatement(TableSpaceManager.java:1144)
	at herddb.core.TableSpaceManager.executeStatementAsyncInternal(TableSpaceManager.java:1082)
	at herddb.core.TableSpaceManager.executeStatementAsync(TableSpaceManager.java:1060)
	at herddb.core.DBManager.executeStatementAsync(DBManager.java:645)
	at herddb.core.DBManager.executePlanAsync(DBManager.java:702)
	at herddb.core.DBManager.executePlan(DBManager.java:672)
	at herddb.core.TestUtils.scan(TestUtils.java:71)
	at herddb.sql.CalcitePlannerTest.joinsPlansTests(CalcitePlannerTest.java:250)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)

What I notice is the Row Expression [RexInputRef] indexes are out of order. 
For query "select k2,n2,n1   from tblspace1.tsql2 join tblspace1.tsql on k2<>k1  where n2 > 1 "
the rexindexes is $3, $4, $1. This causes an exception to be thrown when we try to figure out if the column is primary key. I don't know if this is something new introduced in the latest version of calcite

eolivelli commented:
Is it the only failure?

I think we should adapt to Calcite.
Maybe there is some point in our code in which we are not handling the indexes of RexInputRef correctly.
Or maybe it is a bug on Calcite.

I am not sure if it is better to spend time in adding a testcase in Calcite or debugf

eolivelli commented:
I had clicked 'close and comment' by mistake

amitvc commented:
@eolivelli  I will investigate further. It could be something we are doing. I need to step through the code to see where the disconnect is. I will update you tomorrow.

amitvc commented:
Couple of other assertions that are failing I fixed them. Calcite 1.21 is returning BindableTableScan relnodes and we are expecting TableScan relnodes. Hence some of the asserts were failing. I fixed the UT.

eolivelli commented on herddb-core/src/main/java/herddb/sql/CalcitePlanner.java:
```diff
@@ -883,8 +886,8 @@ private PlannerOp planEnumerableHashJoin(EnumerableHashJoin op, RelDataType rowT
             fieldNames[i] = col.name;
             columns[i++] = col;
         }
-        return new SemiJoinOp(fieldNames, columns,
```
If we are not using this class anymore you can drop it.

amitvc replied:
Ok.

eolivelli commented on herddb-core/src/main/java/herddb/sql/CalcitePlanner.java:
```diff
@@ -884,16 +886,16 @@ private PlannerOp planEnumerableSemiJoin(EnumerableSemiJoin op, RelDataType rowT
-        return new SemiJoinOp(fieldNames, columns,
-                leftKeys, left, rightKeys, right);
+        return new JoinOp(fieldNames, columns,
+                leftKeys, left, rightKeys, right, generateNullsOnLeft, generateNullsOnRight, false);
```
What is 'false'? Can you please add a comment with the name of the argument?
Like:

false /*name of the argument*/

amitvc replied:
Will do.

amitvc commented:
@eolivelli I am getting this error on RawSqlTests::jdbcWrongParameterCountTest when I run the query **SELECT * FROM tblspace1.tsql where n1=? and k1 in (SELECT k1 FROM tblspace1.tsql where n1=?)**

java.lang.RuntimeException: Error while applying rule FilterJoinRule:FilterJoinRule:filter, args [rel#331:EnumerableFilter.ENUMERABLE.[](input=RelSubset#330,condition==($1, ?0)), rel#468:EnumerableHashJoin.ENUMERABLE.[](left=RelSubset#309,right=RelSubset#365,condition==($0, $3),joinType=semi)]

Trying to investigate this now.

eolivelli commented:
@amitvc it is a bug in Calcite.
Please report it to the dev list of Calcite.

For HerdDB it is not a very important case, as it is quite a useless query.

If I remember correctly the test in that point wants to assert that we are able to return an error if the number of jdbc parameters is not correct even in case of a subquery.

You can change the subquery in a way that it still contains a '?' parameter but without breaking Calcite

amitvc commented:
@eolivelli I will report this bug to calcite community and move on to fixing other tests.

eolivelli commented:
Okay.
I would like this test not to be a blocker for our switch.
We can change the query.
I think Calcite is messed up because the subquery is very similar to the main query.

Btw follow your way and move forward in the direction you prefer.
It is an hard task this one, you are on your way!

amitvc commented:
@eolivelli Additional updates. SimpleOperatorsTest::simpleArithmeticOperationsTest test fails for 
**"SELECT * FROM tblspace1.tsql WHERE '1' in (1,2,3)"** query. The reason is BindableTableScan returned by calcite has filters with the following data -
LogicalProject(subset=[rel#11:Subset#2.ENUMERABLE.[]], k1=[$0], n1=[$1], s1=[$2])
  LogicalFilter(subset=[rel#8:Subset#1.NONE.[]], condition=[OR(=(CAST('2'):INTEGER NOT NULL, 1), =(CAST('2'):INTEGER NOT NULL, 2), =(CAST('2'):INTEGER NOT NULL, 3))])
    EnumerableTableScan(subset=[rel#6:Subset#0.ENUMERABLE.[]], table=[[tblspace1, tsql]])

-- Best Plan
EnumerableInterpreter
  BindableTableScan(table=[[tblspace1, tsql]], filters=[[OR(=(CAST('2'):INTEGER NOT NULL, 1), =(CAST('2'):INTEGER NOT NULL, 2), =(CAST('2'):INTEGER NOT NULL, 3))]])

In prior version of calcite this is what we received 
-- Mid Plan
LogicalProject(subset=[rel#11:Subset#2.ENUMERABLE.[]], k1=[$0], n1=[$1], s1=[$2])
  LogicalFilter(subset=[rel#8:Subset#1.NONE.[]], condition=[OR(=('2', CAST(1):CHAR(1) NOT NULL), =('2', CAST(2):CHAR(1) NOT NULL), =('2', CAST(3):CHAR(1) NOT NULL))])
    EnumerableTableScan(subset=[rel#6:Subset#0.ENUMERABLE.[]], table=[[tblspace1, tsql]])

-- Best Plan
EnumerableInterpreter
  BindableTableScan(table=[[tblspace1, tsql]], filters=[[OR(=('2', CAST(1):CHAR(1) NOT NULL), =('2', CAST(2):CHAR(1) NOT NULL), =('2', CAST(3):CHAR(1) NOT NULL))]])

I wonder if the ordering of fields is causing the issue with class casting. I get an error 

java.lang.IllegalArgumentException: Unexpected error on cast of value 1 (class java.lang.String): java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Number

	at herddb.sql.SQLRecordPredicate.cast(SQLRecordPredicate.java:165)
	at herddb.sql.expressions.ConstantExpression.cast(ConstantExpression.java:51)
	at herddb.sql.expressions.SQLExpressionCompiler.compileExpression(SQLExpressionCompiler.java:319)
	at herddb.sql.expressions.SQLExpressionCompiler.compileExpression(SQLExpressionCompiler.java:276)
	at herddb.sql.expressions.SQLExpressionCompiler.compileExpression(SQLExpressionCompiler.java:276)
	at herddb.sql.CalcitePlanner.planBindableTableScan(CalcitePlanner.java:802)
	at herddb.sql.CalcitePlanner.convertRelNode(CalcitePlanner.java:565)
	at herddb.sql.CalcitePlanner.planInterpreter(CalcitePlanner.java:1058)
	at herddb.sql.CalcitePlanner.convertRelNode(CalcitePlanner.java:595)
	at herddb.sql.CalcitePlanner.translate(CalcitePlanner.java:290)
	at herddb.core.TestUtils.scan(TestUtils.java:70)
	at herddb.core.SimpleOperatorsTest.simpleArithmeticOperationsTest(SimpleOperatorsTest.java:298)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Number
	at herddb.sql.SQLRecordPredicate.cast(SQLRecordPredicate.java:155)
	... 33 more

eolivelli commented:
I think you can change the test case, the query does not make sense, as '1' is not a number, it is a string.

What do you think the test is trying to assert? Is it simply a bug in the test?
Maybe we can change '1' in (1,2,3) with 1 in 1,2,3)

eolivelli commented:
So Calcite was adding an implicit cast.
I feel the new behavior is better

amitvc commented:
@eolivelli yes I think so. I am going to adapt this behavior with our code base.

amitvc commented:
@eolivelli In the same UT there is another query like this -
"SELECT * FROM tblspace1.tsql WHERE 'b' in (1)". What are we trying to do here ? Are we saying 'b' is the column index or the actual value? Same with this query "SELECT * FROM tblspace1.tsql WHERE '1' in (1,2,3)"? Are we expecting to cast the value to the appropriate type based on what the type is identified in the set (1,2,3)?

eolivelli commented:
I had the same problem with my pr for the simplification rule.
Change with '1 in (1,2,3)'and similar.
We don't need to support mixed types

amitvc commented:
Ok. Will do. But there are also unit tests that do something like this 
"SELECT * FROM tblspace1.tsql WHERE 'b' in ('1',2,3)"
"SELECT * FROM tblspace1.tsql WHERE '1' in ('1',2,3)"

Are they valid ?

eolivelli commented:
You can fix them, see my other pr, the one you approved yesterday

amitvc commented:
Ok I am able to move forward. All tests are passing in SimpleOperatorsTest.java. 
There is still an issue I reported with Calcite that is blocking some tests. The issue I reported is 
https://issues.apache.org/jira/browse/CALCITE-3347 that is causing these tests to fail
[ERROR]   RawSQLTest.jdbcWrongParameterCountTest:180
[ERROR]   SimpleSubqueryTest.deleteWithSubQueryTest:200
[ERROR]   SimpleSubqueryTest.subQueryOnWhereTest:155

I will ping on calcite dev to see if they need additional information. For now I will comment out these 3 tests and move forward to see if everything else passes.

eolivelli commented:
Sounds like a good plan. Thank you

eolivelli commented:
@amitvc  can you please rebase this branch to latest master ?

eolivelli commented:
@amitvc on my box I have very few test cases failing on this branch, I feel we are close to delivering it

[ERROR] Errors: 
[ERROR]   RawSQLTest.jdbcWrongParameterCountTest:180 » Runtime Error while applying rule...
[ERROR]   SimpleOperatorsTest.simpleArithmeticOperationsTest:298 » IllegalArgument Unexp...
[ERROR]   SimpleSubqueryTest.deleteWithSubQueryTest:200 » Runtime Error while applying r...
[ERROR]   SimpleSubqueryTest.subQueryOnWhereTest:155 » Runtime Error while applying rule...
[INFO] 
[ERROR] Tests run: 418, Failures: 0, Errors: 4, Skipped: 2

amitvc commented:
@eolivelli I pushed the changes for broken UT and all the UT's in herddb pass now. I think the ones i removed are dependent on the fix from calcite.

amitvc commented:
@eolivelli I will rebase, resolve the conflicts and push the changes tonight.

eolivelli commented:
Great

amitvc commented:
@eolivelli All tests pass now and i have rebased with master.

eolivelli commented on herddb-core/src/main/java/herddb/sql/CalciteEnumUtils.java:
```diff
@@ -0,0 +1,26 @@
+            case ANTI:
+                return JoinType.ANTI;
+        }
+        throw new IllegalStateException(
```
move this to a 'default' case

amitvc commented:
@eolivelli checkstyle: check should be all green now.

eolivelli commented:
Cool.
Github seems broken currently, it tells me 'server error'
I will merge as soon as it is working again.

Thank you so much @amitvc

We should follow our issues on Calcite and upgrade as soon as possible.

Please create an issue on our github with a link to the fixes that we are waiting from Calcite community

eolivelli commented:
@amitvc can you create a new Pull request with this work, targeting Calcite 1.22.0-SNAPSHOT ?
please uncomment the failing tests about subqueries with bind variables, now they should pass

amitvc commented:
@eolivelli will do.
